### DISCUSSION TITLE
Is it safe to destroy an object with a running tween on it?

### DISCUSSION THREAD

**Discussion by KyryloKuzyk** [KyryloKuzyk/PrimeTween#4] [Announcements]
Is it safe to destroy an object with a running tween on it?
---

The short answer is **yes**! It's safe to destroy an object with a running tween. Suppose the tween's target is UnityEngine.Object, then PrimeTween will check if it's not destroyed before playing an animation further. This means you shouldn't care if an object has running tweens before destroying it or before unloading a scene.

But if a tween has an `OnComplete()` callback, and the tween's target gets destroyed, then what?

#### 'warnIfTargetDestroyed' parameter
PrimeTween never silences errors or potential places where a bug can occur in a user's code. So the solution to this problem is simple - **log an error** that OnComplete() will not be called because of the interruption. This attracts the user's attention to the potential problem and lets them fix it correctly.

So the rule of thumb is to NOT use the OnComplete() for **important actions** if you **expect the tween's target to be destroyed mid-way**. If you're using OnComplete() only for cosmetics, then pass the `warnIfTargetDestroyed: false` parameter to the OnComplete() method to silence PrimeTween's warning.
```csharp
Tween.PositionX(transform, endValue, duration)
    .OnComplete(transform, _transform => {
        // Cosmetic code that can be safely ignored if the transform is destroyed mid-way.
        _transform.localScale = Vector3.zero;
    }, warnIfTargetDestroyed: false);
```

#### Coin indicator example.
Let's consider the next piece of code. It animates the coinIndicator, then gives coins to the player, and updates the label:
```csharp
Tween.Scale(coinIndicator, 1, 1.5f, duration, cycles: 2, cycleMode: CycleMode.Rewind)
    .OnComplete(() => {
        AddCoins(100); // if coinIndicator is destroyed mid-way, the AddCoins(100) will not be called, and PrimeTween will warn you about that
        UpdateCoinIndicatorLabel();
    });
```
The issue with the above code is that if the coinIndicator is destroyed (or the scene is unloaded) before the animation finishes, the AddCoins(100) method will not be called, and the player will not receive their hard-earned coins. If this happens, PrimeTween will warn about that with an error to the console.

Here is a fixed version that executes only cosmetics code in the OnComplete() callback:
```csharp
AddCoins(100); // Call the important method independently from the animation
Tween.Scale(coinIndicator, 1, 1.5f, duration, cycles: 2, cycleMode: CycleMode.Rewind)
    .OnComplete(() => {
        UpdateCoinIndicatorLabel(); // OnComplete() callback executes only cosmetic code that can be safely ignored if coinIndicator is destroyed
    }, warnIfTargetDestroyed: false); // disable PrimeTween's warning with 'warnIfTargetDestroyed: false' 
```
Notice that the important AddCoins(100) method is called independently from the animation, and the OnComplete() method only updates the label. The `warnIfTargetDestroyed` is set to false to tell PrimeTween that the callback is safe to ignore if coinIndicator gets destroyed mid-way.

#### Why not add the 'OnTargetDestroyed()' callback to the tween?
If you want to do some kind of clean-up when the tween's target gets destroyed, then such a callback will not solve the problem, but will only make it worse. Let's consider a hypothetical example:
```csharp
// OnTargetDestroyed() is NOT present in PrimeTween; this is just a hypothetical example
Tween.PositionX(transform, endValue, duration)
    .OnTargetDestroyed(() => DoCleanUp()); // Warning! A lot can go wrong inside the OnTargetDestroyed() callback!
// ...
Object.Destroy(gameObject);
```
After the transform is destroyed, it's most likely too late to do the clean-ups: (1) you can't do any operations with the transform; (2) you can't do any operations with the MonoBehaviour that started the tween because the references it holds may already be destroyed too.

The correct way to address these kinds of problems is to do a clean-up manually **before** you destroy the target. This way, you'll not only prevent all kinds of MissingReferenceException but will also make your code more readable and reliable.
```csharp
Tween.PositionX(transform, endValue, duration);
// ...
DoCleanUp();
Object.Destroy(gameObject);
```

#### Global PrimeTweenConfig.warnIfTargetDestroyed setting
In version 1.3.8, I added a new global **`PrimeTweenConfig.warnIfTargetDestroyed`** setting, which lets you disable the warning globally. I still recommend against it because while it adds convenience, it can eventually silence a real problem, which can waste hours of debugging. Use this setting only if all animations in your project are cosmetic and their `OnComplete` callbacks never have business logic.

**Comment by Wokarol** [comment]
What would be the correct approach to do a coin animation that adds the coins once the animation finished in a situation where the level could finish before the end of the animation?
(For the sake of the discussion let's assume that the player's wallet is persistent because of DontDestroyOnLoad)

The solution of moving the important method outside the callback doesn't really work in this scenario as it changes behaviour.

We also cannot ignore that as a cosmetic change because then the player will lose those coins.

**Reply by KyryloKuzyk** [reply]
> What would be the correct approach to do a coin animation that adds the coins once the animation finished in a situation where the level could finish before the end of the animation?

I'm not a proponent of the VMC pattern, but I'll use it to draw the analogy. Counting coins is a 'model', and it should be consistent, correct, and always up-to-date. Animating the coin indicator and displaying the coin's label is the 'controller' part that should not influence the 'model'. I'm not advocating for an MVC pattern, but I believe that separating the data from animations is a good decision.

However, if you still want to actually add the coins once the animation completes, it's possible by using a separate Tween.Delay() call. By adding coins in a separate Tween.Delay() you decouple important code from the animation (because the animation can potentially be interrupted by the scene change or target destruction). And the target-less Tween.Delay() will survive the coinIndicator destruction or scene unload.

```csharp
public void AddCoinsWithAnimation() {
    Tween.Scale(coinIndicator, 1, 1.5f, duration, cycles: 2, cycleMode: CycleMode.Rewind)
        .OnComplete(() => UpdateCoinIndicatorLabel(), warnIfTargetDestroyed: false);
    Tween.Delay(duration, () => PersistentPlayerWallet.Instance.AddCoins(100));
}
```

However, the above code still has a problem: if the application terminates before the Tween.Delay() is completed, then the player will not receive the coins. So the completely correct solution is to still count coins separately from the animation.

> The solution of moving the important method outside the callback doesn't really work in this scenario as it changes behaviour.

Yes, correctly, it changes behavior, but it fixes the potential bug. Adding important code to the OnComplete() is a bug because the animation can be interrupted by a target destruction or a scene change (or the app termination). OnComplete() is called when the animation **completes** (i.e. runs to completion), but scene unload and target destruction **interrupt** the animation, preventing its completion.

**Comment by NFMynster** [comment]
Would love for this to also be configurable globally, ie `PrimeTweenConfig.warnIfTargetDestroyed `

**Reply by kiroZhang1997** [reply]
I agree with you

**Reply by KyryloKuzyk** [reply]
@NFMynster @kiroZhang1997 
> Would love for this to also be configurable globally, ie `PrimeTweenConfig.warnIfTargetDestroyed`

This setting can easily be added. However, I have concerns that it will open the possibility for users to shoot themselves in a foot in some cases.

My idea with the 'warnIfTargetDestroyed' parameter was that users should make a decision on a case-by-case basis to distinguish 'cosmetic' callbacks from those that are critical and can't be safely ignored. 

With the global `PrimeTweenConfig.warnIfTargetDestroyed` setting, it would be very easy to forget about the global setting and make a mistake. Every OnComplete() callback is different, and deciding explicitly which 'warnIfTargetDestroyed' parameter to use for each OnComplete() callback is more reliable. On the other hand, setting global `PrimeTweenConfig.warnIfTargetDestroyed = false`, will disable all warnings in your project (and more importantly, in all future code you will write), so you can easily miss a potential problem in your code.

What do you think?

**Reply by NFMynster** [reply]
Now yes, it is a dangerous path for sure, I agree there, though I personally don't use tweens for anything business critical - and if I do, I have my tweens stored as variable for better control.

I would maybe always have the warning enabled if editor, but not in release builds, which this flag would help with.
But I get your argumentation and won't really counter argue - it's valid points which I've considered myself. "Do I really want this?"

**Comment by NFMynster** [comment]
Is Sequence.ChainDelay missing the `warnIfTargetDestroyed` on purpose? Or does it not matter?

**Reply by NFMynster** [reply]
Ok, after using this more and more, it's a consistent issue. I would much rather have the flag the other way around, so if you have a business critical tween, you set warnIfTargetDestroyed: true, explicitly.

**Reply by NFMynster** [reply]
If not, I'd like a Tween.StopAll (already there) but also a sequence.stopAll

**Comment by KyryloKuzyk** [comment]
> Is Sequence.ChainDelay missing the warnIfTargetDestroyed on purpose? Or does it not matter?

It doesn't matter because Sequence.ChainDelay() doesn't have the callback. 'warnIfTargetDestroyed' is only present in methods that have callbacks.
> I would much rather have the flag the other way around, so if you have a business critical tween, you set warnIfTargetDestroyed: true, explicitly.

Changing the default value of 'warnIfTargetDestroyed' from true to false will make it a lot harder for users to debug problems related to destroying objects. Oftentimes, users new to Unity and C# are not aware of this potential problem at all. So I decided to go on the safe path and treat all OnComplete() callbacks as 'business critical' by default, leaving it to users to decide on a case-by-case basis if it's safe to ignore the particular callback or not.
Can you please tell me how often you encounter the need to set `warnIfTargetDestroyed = false`? If you don't expect the object to be destroyed while the animation is still running, you can omit the warnIfTargetDestroyed and leave it at the default 'true' value. Or if your game can potentially change scenes at arbitrary moments, you can call Tween.StopAll() before changing scenes.

> If not, I'd like a Tween.StopAll (already there) but also a sequence.stopAll

Tween.StopAll() stops all tweens and sequences. Or do you wish to have a separate Sequence.StopAll() method that will stop only sequences?

**Reply by NFMynster** [reply]
Yes, after writing this I did do the `Tween.StopAll()`, but I worried that it didn't stop sequences. It seems to be the case, fortunately.
> Or if your game can potentially change scenes at arbitrary moments,

This is exactly my case, I'm developing a multiplayer game where the host (or client) can disconnect at any moment. I did end up just doing Tween.StopAll when disconnection happens, and this solved all my issues.
But the XML summary of Tween.StopAll made me worry that it didn't stop the sequences also.

So, thank you first of all, I rest my case regarding defaulting `warnIfTargetDestroyed` to false - my case is solved by manually killing all my tweens. 
--This wasn't viable when I used DOTween, due I had a bunch of references to sequences and such which I played forwards/backwards, and killing them all broke that, however, now it is doable.

I am also happy to share that I've ported my entire game to PrimeTween and deployed it, everything is working great.

**Reply by KyryloKuzyk** [reply]
I'm glad to hear that PrimeTween works perfectly for your game and even solves limitations present before with other tween libraries!  
I'll update the XML docs in the next version to make it clear that Tween.StopAll/CompleteAll() also include all sequences.

**Comment by YuliyF** [comment]
```
Tween.Delay(gameObject, 4f, () => ExitToMain(), warnIfTargetDestroyed: false);
void ExitToMain()
{
 ...
Tween.StopAll(gameObject);
Tween.StopAll();
}
```
For me It doesn't work, I catch an error

**Reply by KyryloKuzyk** [reply]
I don't have the complete picture, but I assume you destroy the current scene at the beginning of the ExitToMain() method, right? If this is the case, the 'Tween's OnComplete callback was ignored' warning will come not from the Delay tween, but from some other tween whose target has been destroyed by the scene unload. So you should use the 'warnIfTargetDestroyed' parameter on that **other tween**, not on the Delay itself.

Also, unloading the scene **after** the Tween.StopAll() will also solve the issue without the need to use the 'warnIfTargetDestroyed' parameter at all:
```csharp
void ExitToMainAfterDelay() {
    Tween.Delay(gameObject, 4f, () => ExitToMain()); // warnIfTargetDestroyed: false is not needed here if you don't destroy the 'gameObject' before the Delay completes 
}

void ExitToMain() {
    // Tween.StopAll(gameObject); // this call is not needed because the Delay tween has been already completed and we're currently in its onComplete callback 
    Tween.StopAll();
    UnloadCurrentScene(); // unload the current scene AFTER Tween.StopAll() to prevent the 'Tween's OnComplete callback was ignored' warning from destroyed tweens
}
```

**Comment by Nomad1108** [comment]
One thing I've noticed is that if a tween isPaused = true when its gameObject target is destroyed, the tween will persist with Unity Target set to "Missing (Object)" in the Inspector. 

Is it the expected behavior? My assumption was that destroying its target would kill the tween...

<img width="652" alt="Capture d’écran 2024-09-05 à 11 17 36" src="https://github.com/user-attachments/assets/81348a3a-d98f-4de9-96a5-d69933af827b">


So I fixed this by adding this to the OnDisable() method of the target:

```
if (_tween.isAlive)
    _tween.Stop();
```

**Reply by KyryloKuzyk** [reply]
That’s a great find, thank you!

Currently, PrimeTween doesn’t check for the destroyed object when the tween is paused. I’ll fix that in the next update.

**Reply by KyryloKuzyk** [reply]
I fixed this in version 1.1.22. 
https://github.com/KyryloKuzyk/PrimeTween/blob/main/changelog.md
Huge thanks for reporting the problem!

**Reply by Nomad1108** [reply]
Very happy I could contribute to such an awesome asset!

**Comment by kreso22** [comment]
It would be nice to have a global opt-out.

My game simulation (game logic) is completely separate from view (tweens etc).
Tweens have no effect on what happens in the game.
There is no benefit in having this error pop up in my situation. At best it could be an opt-in warning. Error make no sense.

**Reply by KyryloKuzyk** [reply]
@kreso22 I totally understand where your request is coming from. This post addresses exactly that: https://github.com/KyryloKuzyk/PrimeTween/discussions/4#discussioncomment-7351778

You can either 1) set `warnIfTargetDestroyed: false`, or 2) stop the animation manually before destroying the target. If you see the error during a scene change, you can also call `Tween.StopAll / CompleteAll` before loading a new scene.

**Reply by kreso22** [reply]
Thanks for your response @KyryloKuzyk I understand the rationale, but I disagree with the approach.

Having to set `warnIfTargetDestroyed: false` individually on every tween - or add manual cleanup logic everywhere - creates significant boilerplate for a non-issue in many codebases. My simulation layer is completely decoupled from the view layer. Tweens are purely cosmetic. There's no scenario where a missed callback affects game state.

The argument against a global flag seems to be "users might enable it without understanding the consequences." But that logic could apply to almost any configuration option. Most libraries trust developers to read documentation and make informed choices about their own projects.

A global opt-out doesn't prevent anyone from using the per-tween flag when they need fine-grained control. It just respects that some of us have already made an architectural decision that makes this warning noise rather than signal.

I'd still appreciate this being reconsidered.

**Reply by JoanOrtiga-AntiHero** [reply]
It's insane to me that this is not an option. Give people options, specially on a tweening library where many ppl use it as a view only, not for logic. The fact that I cannot disable on release builds warnings is making me re-consider the use of this tweening library.

Give me the concious choice to disable this warnings, AT LEAST ON RELEASE.

**Comment by KyryloKuzyk** [comment]
@kreso22 @JoanOrtiga-AntiHero @NFMynster @Wokarol 
Good news! In version 1.3.8, I added a new global `PrimeTweenConfig.warnIfTargetDestroyed` setting, which lets you disable the warning globally.

I still recommend against it because while it adds convenience, it can eventually silence a real problem, which can waste hours of debugging.

**Reply by JoanOrtiga-AntiHero** [reply]
Thank you for reconsidering it. My intent has been always adding it only to release builds, so a missed tween creates logs on player's device. I'll keep it on on development builds while testing to realize if needed. Thank you.

**Comment by Wokarol** [comment]
@KyryloKuzyk // As a way to avoid hidden errors. What about borrowing the naming from C# try...catch.
     
What I am suggesting is creating two callbacks. OnComplete and OnFinalize.    
OnFinalize is called always when the tween stops playing for whatever reason (destruction of the object being one of them)
OnComplete gets called when the tween completes.

That would allow us to pick a method based on the intended use case. Finalize would be used for logic, such as "add those points with a delay" while Complete would be used for purely tween related visual things like disabling an object after the tween ends.

It would mimick the try...catch...finalize C# users are familiar with (I hope)

**Comment by Wokarol** [comment]
The added benefit is that if Finalize callback accesses a destroyed object, it would naturally raise an exception

**Comment by KyryloKuzyk** [comment]
@Wokarol 
> What I am suggesting is creating two callbacks. OnComplete and OnFinalize.
> OnFinalize is called always when the tween stops playing for whatever reason (destruction of the object being one of them)
> The added benefit is that if Finalize callback accesses a destroyed object, it would naturally raise an exception

Stopping an animation via `tween.Stop()` and destroying the target are two completely different reasons for stopping the animation. The first one is the manual stop, and the other one is an exceptional stop. It's safe to access the target in the first case, but not in the second.

Combining them under one `OnFinalize` API just doesn't feel right. I'd like for PrimeTween users to differentiate between the two scenarios.
I also touched on this topic in the `Why not add the 'OnTargetDestroyed()' callback to the tween?` section. 

> The added benefit is that if Finalize callback accesses a destroyed object, it would naturally raise an exception

When Unity unloads a scene, all objects are destroyed in an undeterministic order, so such an exception will only happen occasionally, which will complicate its fix. Instead, I prefer not to allow situations that can throw exceptions in my users' projects.

**Comment by Wokarol** [comment]
> I'd like for PrimeTween users to differentiate between the two scenarios.

You would still be able to do that tho. Since you end up with two callbacks. One that runs when the tween stop/completes normally. And the other than will always run no matter what the reason for the stop was. Again, this is a pattern already used for exception handling in C#.

> all objects are destroyed in an undeterministic order, so such an exception will only happen occasionally

Would that actually happen occasionally? Yes, the order at which the tweens would stop would be random. But OnFinalize would be called always after the target was destroyed. So if a callback accesses the target in na unsafe manner, it would always lead to an exception

**Reply by KyryloKuzyk** [reply]
@Wokarol  I'm not a fan of exceptions in general, as they add unnecessary complexity to programs and obscure the program flow. Destroying an object with an animation is not an "exceptional" situation, but a completely normal program flow. I don't want my users to pollute their code with try-catches, especially when there is a cleaner, more straightforward approach.

**Reply by Wokarol** [reply]
The issue is that I wouldn't call the current approach that much cleaner. Let take a practical example. A coin flying towards the player. 

We create a simple animation using a tween that makes it so the coin upon being picked up flies towards the player. And with a delay, adds the value to their wallet.
To make this effect nicer, I would want to add the money after the coin finished the flight. I can do that via a callback reacting to a completion of the tween.

It's all and nice... until the player changes the level before the coin reached them. I would want the money to be counted always, even if the coin was destroyed due to an unload.

How would you solve that? With OnFinalize, that would be simple and clean.

**Reply by kreso22** [reply]
This is a great idea! Different behavior = different outcome, via it's own corresponding method.
